Reduce HAProxy reloads - adds support to use the haproxy dynamic config api

ramr opened the pull request:
Add support to the template router to dynamically configure haproxy. This reduces the reloads 
needed when the router configuration changes due to routes/endpoints changing.

Note that HAProxy doesn't have support to dynamically change certificates, so we need to pre-allocate pools of "blueprint" routes to dynamically bring new routes online.

* We now add/remove servers from a backend on endpoint changes (for scale up/down, service addition/removal/update, node drain/evictions etc).
   Each haproxy backend (for a route) has a set of dynamic servers added named 
   `_dynamic-pod-[1-5]` - which are used when endpoints change (example for scale up).
    Scaling down endpoints just sets the existing servers in a haproxy backend to "maint" or
    non-serving state.

   The infra router options - command line flags or environment variables named  
   `ROUTER_DYNAMIC_SERVER_PREFIX` and `ROUTER_MAX_DYNAMIC_SERVERS` (default: 5) 
   controls these options.

* We dynamically add and remove routes if a route matches a pre-allocated/configured blueprint route.
   The default config has blueprint routes for insecure/http, edge terminated and passthrough routes.
   Re-encrypt routes are an issue as they could have custom certificates so is not currently in that 
   default blueprint set (commented out). 
   Route removal just removes the map entries, marks the servers in `maint` state but leaves the 
   backend config in haproxy intact. The backend is not referenced from anywhere, so is unused.

    Route addition checks if it can bring the route online without a router reload - it searches for a
    matching blueprint and then a free slot from the pre-configured pool of  routes for that blueprint.
    If the new route can't be dynamically configured, then the behavior is as today and a reload occurs.

    A route might not match a blueprint if it has information that causes a custom haproxy config.  
   Examples of that could be a route that has custom certificate information or custom CA information   
   to validate re-encrypted routes or it could even be for custom config that is generated for the 
   route annotations. Those would fail to find a matching blueprint (and hence cause a reload).

    However that said, the template router does also provide the ability to add custom blueprints. 
    These are basically just routes in a namespace which serve as templates or blueprints - similar
    to the default blueprints (for insecure/http, edge terminated and passthrough routes). 
    So one can easily add a re-encrypt route blueprint if needed. 
    Also, the add route code would search the custom blueprints (in the same fashion that it does
    the default blueprints - custom ones are just added to the list). 

    The infra router options 
        `ROUTER_BLUEPRINT_ROUTE_NAMESPACE` and 
        `ROUTER_BLUEPRINT_ROUTE_POOL_SIZE` (default: 10) 
   provide some knobs for these custom blueprints.

  Tests will automatically use this new functionality (extended ones needed a config change) but the
  others use the router image, so automatically check with the haproxy router with the dynamic config
  manager enabled.
   
   Again, a point to note here is that if we exhaust the pool of dynamic servers or blueprint routes or 
   don't find a matching blueprint, the template router will need/do a reload.

Associated trello card: https://trello.com/c/W21FD0v0/585-13-implement-dynamic-changes-to-the-router-routerscaledemo

/cc @knobunc  @zhaozhanqi fyi

smarterclayton commented on glide.yaml:
```diff
@@ -6,181 +6,133 @@ excludeDirs:
 - package: k8s.io/apiserver
-  repo:    git@github.com:openshift/kubernetes-apiserver
   version: release-1.9.1
+  repo: git@github.com:openshift/kubernetes-apiserver
```
Please undo this reformatting.

ramr replied:
That's all done by glide. I couldn't run glide on my local repo - it doesn't run for me, times out all the time, so @rajatchopra  ran it for me  and he didn't do anything other than `glide get` && `./hack/update-deps.sh`.  Is something that's glide version specific causing this?

smarterclayton replied:
I’ve never had glide time out - are you on an old version?  Bad internet?

ramr replied:
glide version is v0.13.1  and its a fast[ish] pipe - 200mb+ consistently (and also tried it on a T3 connection) - so i suspect its not the speed. Plus did try it on 3 different vms as well (all fedora 27 though), so something else. Anyway, will look at it at some other time.

smarterclayton commented on vendor/k8s.io/kubernetes/pkg/volume/configmap/configmap.go:
```diff
@@ -194,9 +194,6 @@ func (b *configMapVolumeMounter) SetUpAt(dir string, fsGroup *int64) error {
 	if err := wrapped.SetUpAt(dir, fsGroup); err != nil {
 		return err
 	}
-	if err := volumeutil.MakeNestedMountpoints(b.volName, dir, b.pod); err != nil {
```
Please fix this, you reverted changes.

ramr replied:
Hmm, maybe because we did the `glide get` and `update-deps.sh` a few weeks back - this might be outdated. Will do.

smarterclayton commented on test/end-to-end/router_test.go:
```diff
@@ -1324,6 +1324,10 @@ func createAndStartRouterContainerExtended(dockerCli *dockerClient.Client, maste
 		fmt.Sprintf("NAMESPACE_LABELS=%s", namespaceLabels),
+		fmt.Sprintf("ROUTER_CONFIG_MANAGER=haproxy-manager"),
+		fmt.Sprintf("ROUTER_DYNAMIC_SERVER_PREFIX=_test-dynamic"),
+		fmt.Sprintf("ROUTER_MAX_DYNAMIC_SERVERS=3"),
```
These should have reasonable defaults.  I'd prefer if your testing required only ROUTER_CONFIG_MANAGER unless there's a good reason not to.

ramr replied:
ok - i'll set it to only the defaults.

smarterclayton commented on pkg/router/template/types.go:
```diff
@@ -136,6 +140,88 @@ type certificateWriter interface {
+	Initialize(router RouterInterface, certPath string)
+
+	// Register registers an id to be associated with a route.
+	Register(id string, route *routeapi.Route, wildcard bool)
```
Why are you passing wildcard here when it should be on the route?

ramr replied:
We already had boolean deciphered based on the policy so I just passed the boolean but yeah we could `recompute` it in register/add.

smarterclayton commented on pkg/router/template/types.go:
```diff
@@ -136,6 +140,88 @@ type certificateWriter interface {
+
+// ConfigManager is used by the router to make configuration changes using
+// the template router's dynamic configuration API (if any).
+type ConfigManager interface {
```
Please specify the threading guarantees this code requires with the router controller here.

ramr replied:
ack

smarterclayton commented on pkg/router/template/configmanager/haproxy/map.go:
```diff
@@ -0,0 +1,209 @@
+package haproxy
```
Give this its own unit tests.

ramr replied:
there's a `map_test.go` in there. Or did you mean something else?

smarterclayton commented on pkg/router/template/configmanager/haproxy/converter.go:
```diff
@@ -0,0 +1,88 @@
+}
+
+// NewCSVConverter returns a new CSVConverter.
+func NewCSVConverter(headers string, out interface{}, fn ByteConverterFunc) CSVConverter {
```
Should be a pointer unless you have a good reason not to.

ramr replied:
Not really - ok will change.

smarterclayton commented on pkg/router/template/configmanager/haproxy/converter.go:
```diff
@@ -0,0 +1,88 @@
+
+// Convert runs a haproxy dynamic config API command.
+func (c CSVConverter) Convert(data []byte) ([]byte, error) {
+	glog.V(4).Infof("CSV converter input data bytes: %s", string(data))
```
How much data is this?  If it's going to regularly be more than a few hundred bytes it should be at v5.

ramr replied:
yeah it might be more than a few hundred - will set to v5.

smarterclayton commented on pkg/router/template/configmanager/haproxy/converter.go:
```diff
@@ -0,0 +1,88 @@
+	if c.converterFunc != nil {
+		convertedBytes, err := c.converterFunc(data)
+		if err != nil {
+			glog.Errorf("CSV converter error: %v", err)
```
Don't log errors like this, wrap or log when you eat it with utilruntime.HandleError

ramr replied:
ok will fix.

ramr replied:
I think this was a remnant from the logs I was generating during dev - will remove this as its not needed - error are anyways propagated up and handled by the caller.

smarterclayton commented on pkg/router/template/configmanager/haproxy/converter.go:
```diff
@@ -0,0 +1,88 @@
+// fixupHeaders fixes up haproxy API responses that don't contain any CSV
+// header information. This allows us to easily parse the data and marshal
+// into an array of native golang structs.
+func (c CSVConverter) fixupHeaders(data []byte) ([]byte, error) {
```
don't make this a method, just a function that takes headers and data.

Give this file unit tests.

ramr replied:
will do

smarterclayton commented on pkg/router/template/configmanager/haproxy/client.go:
```diff
@@ -0,0 +1,204 @@
+}
+
+// isRetryable checks if a haproxy command can be retried.
+func isRetryable(err error, cmd string) bool {
```
`isRetriable`

smarterclayton commented on pkg/router/template/configmanager/haproxy/client.go:
```diff
@@ -0,0 +1,204 @@
+// Backends returns the list of configured haproxy backends.
+func (c *Client) Backends() ([]*Backend, error) {
+	if len(c.backends) == 0 {
+		if backends, err := GetHAProxyBackends(c); err != nil {
```
If these methods don't need to be public (`GetHAProxyBackends`) they shouldn't be public.

smarterclayton commented on pkg/router/template/configmanager/haproxy/client.go:
```diff
@@ -0,0 +1,204 @@
+// if the error for the command is a retryable error.
+func (c *Client) runCommandWithRetries(cmd string, limit int) (*bytes.Buffer, error) {
+	retryAttempt := 0
+	for {
```
use util.ExponentialBackoff or util.PollImmediate instead of inventing your own code for it.

smarterclayton commented on pkg/oc/admin/router/router.go:
```diff
@@ -684,11 +684,15 @@ func RunCmdRouter(f *clientcmd.Factory, cmd *cobra.Command, out, errout io.Write
-		env["ROUTER_METRICS_TLS_CERT_FILE"] = "/etc/pki/tls/metrics/tls.crt"
-		env["ROUTER_METRICS_TLS_KEY_FILE"] = "/etc/pki/tls/metrics/tls.key"
+	if cfg.Type == "haproxy-router" {
+		env["ROUTER_CONFIG_MANAGER"] = "haproxy-manager"
```
If everyone is going to get these just make these the new default on the command.  Don't force people to change their config and make a pointless choice.  Otherwise you'll have to add this to the upgrade scripts.

ramr replied:
ok will do.

ramr replied:
hmm, so because the template router is used for f5 (going away but still) as well as nginx, this needs to be either set here since its haproxy specific or we need to pass down the type of router to the infra command (that's not done today). I can either pass down this new setting or then the type and set the defaults based on that in the infra command ... but in either case it still means upgrading existing haproxy-router deployments. So which do you prefer?

smarterclayton replied:
I think if the config manager works, there’s no reason we wouldn’t have it on by default for haproxy.  However, we will likely default it off at first until we’re comfortable that we’ve caught all the edge cases and then default it on.  So ideally we just have a Boolean that has an internal default and if the user specifies it we obey their rule.  That avoids end users ending up with the flag in their config.

smarterclayton commented on pkg/cmd/infra/router/template.go:
```diff
@@ -438,3 +492,21 @@ func (o *TemplateRouterOptions) Run() error {
+// blueprintRoutes returns all the routes in the blueprint namespace.
+func (o *TemplateRouterOptions) blueprintRoutes(routeclient *routeinternalclientset.Clientset) ([]*routeapi.Route, error) {
+	blueprints := make([]*routeapi.Route, 0)
+	if len(o.BlueprintRouteNamespace) == 0 {
```
I don't like an entire namespace of route blueprints.  Require them to be identified with a label or associated metadata within a specific namespace, or require them to be specified by name directly.

Whole namespaces devoted to this is pretty ugly, although the idea of pulling it from the route is good.

ramr replied:
ok will do.

smarterclayton commented on pkg/cmd/infra/router/template.go:
```diff
@@ -371,6 +401,42 @@ func (o *TemplateRouterOptions) Run() error {
+
+	var cfgManager templateplugin.ConfigManager
+	if o.ConfigManagerName == "haproxy-manager" {
+		blueprintRoutes, err := o.blueprintRoutes(routeclient)
```
You're going to have to refresh these, they shouldn't be static or require restart.  Make that dynamic (add a controller plugin wrapper that triggers reload of the blueprints).

ramr replied:
There's going to be some issues with blueprint refresh that I need to think about - basically vis-a-vis replacement and deletion and how the allocations/config changes, so can we do this in a follow-up PR?

smarterclayton commented on pkg/cmd/infra/router/template.go:
```diff
@@ -113,12 +132,19 @@ func (o *TemplateRouter) Bind(flag *pflag.FlagSet) {
+	flag.DurationVar(&o.CommitInterval, "commit-interval", getIntervalFromEnv("COMMIT_INTERVAL", defaultCommitInterval), "Controls how often to commit (to the actual config) all the changes made using the router specific dynamic configuration manager.")
+	flag.StringVar(&o.BlueprintRouteNamespace, "blueprint-route-namespace", util.Env("ROUTER_BLUEPRINT_ROUTE_NAMESPACE", ""), "Specifies the namespace which contains the routes that serve as blueprints for the dynamic configuration manager.")
+	flag.IntVar(&o.BlueprintRoutePoolSize, "blueprint-route-pool-size", int(util.EnvInt("ROUTER_BLUEPRINT_ROUTE_POOL_SIZE", 10, 1)), "Specifies the size of the pre-allocated pool for each route blueprint managed by the router specific dynamic configuration manager. This can be overriden by an annotation router.openshift.io/pool-size on an individual route.")
+	flag.StringVar(&o.DynamicServerPrefix, "dynamic-server-prefix", util.Env("ROUTER_DYNAMIC_SERVER_PREFIX", ""), "Specifies the prefix for dynamic servers added to router backends. These dynamic servers are handled by the router specific dynamic configuration manager.")
```
I don't see why this is configurable.  Seems pointless?

ramr replied:
Removed this.

smarterclayton commented on pkg/cmd/infra/router/template.go:
```diff
@@ -113,12 +132,19 @@ func (o *TemplateRouter) Bind(flag *pflag.FlagSet) {
 	flag.StringVar(&o.Ciphers, "ciphers", util.Env("ROUTER_CIPHERS", ""), "Specifies the cipher suites to use. You can choose a predefined cipher set ('modern', 'intermediate', or 'old') or specify exact cipher suites by passing a : separated list.")
 	flag.BoolVar(&o.StrictSNI, "strict-sni", isTrue(util.Env("ROUTER_STRICT_SNI", "")), "Use strict-sni bind processing (do not use default cert).")
 	flag.StringVar(&o.MetricsType, "metrics-type", util.Env("ROUTER_METRICS_TYPE", ""), "Specifies the type of metrics to gather. Supports 'haproxy'.")
+	flag.StringVar(&o.ConfigManagerName, "config-manager", util.Env("ROUTER_CONFIG_MANAGER", ""), "Specifies the manager to use for dynamically configuring changes with the underlying router. Supports 'haproxy-manager'.")
```
I don't see a reason not to default this on.

ramr replied:
So we don't pass the router type to the infra command, so this needs to be set from above as of now. We can't set this for nginx template routers. We can either pass the type of router down (ala haproxy etc) and then do the required defaults here or then (as was done) pass this env variable down.

smarterclayton replied:
Do the defaulting later when you know the type.  I doubt we’ll ever have multiple, so this is really just a bool.

smarterclayton commented on pkg/cmd/infra/router/template.go:
```diff
@@ -113,12 +132,19 @@ func (o *TemplateRouter) Bind(flag *pflag.FlagSet) {
 	flag.BoolVar(&o.StrictSNI, "strict-sni", isTrue(util.Env("ROUTER_STRICT_SNI", "")), "Use strict-sni bind processing (do not use default cert).")
 	flag.StringVar(&o.MetricsType, "metrics-type", util.Env("ROUTER_METRICS_TYPE", ""), "Specifies the type of metrics to gather. Supports 'haproxy'.")
+	flag.StringVar(&o.ConfigManagerName, "config-manager", util.Env("ROUTER_CONFIG_MANAGER", ""), "Specifies the manager to use for dynamically configuring changes with the underlying router. Supports 'haproxy-manager'.")
+	flag.StringVar(&o.ConfigManagerConnectionInfo, "config-manager-connection-info", "", "Specifies connection information for the dynamic configuration manager.")
```
Seems overkill to configure this.  We don't allow configuring the stats socket.

ramr replied:
Well the stats socket is sorta tied into how the template is configured (using af unix sockets). 
It probably is overkill but if the haproxy template is potentially changed by the user, this is just a knob to use tcp if so needed.

ramr replied:
I'll remove this - we can add it later if needed.

smarterclayton commented on pkg/cmd/infra/router/template.go:
```diff
@@ -113,12 +132,19 @@ func (o *TemplateRouter) Bind(flag *pflag.FlagSet) {
+	flag.StringVar(&o.ConfigManagerConnectionInfo, "config-manager-connection-info", "", "Specifies connection information for the dynamic configuration manager.")
+	flag.DurationVar(&o.CommitInterval, "commit-interval", getIntervalFromEnv("COMMIT_INTERVAL", defaultCommitInterval), "Controls how often to commit (to the actual config) all the changes made using the router specific dynamic configuration manager.")
+	flag.StringVar(&o.BlueprintRouteNamespace, "blueprint-route-namespace", util.Env("ROUTER_BLUEPRINT_ROUTE_NAMESPACE", ""), "Specifies the namespace which contains the routes that serve as blueprints for the dynamic configuration manager.")
+	flag.IntVar(&o.BlueprintRoutePoolSize, "blueprint-route-pool-size", int(util.EnvInt("ROUTER_BLUEPRINT_ROUTE_POOL_SIZE", 10, 1)), "Specifies the size of the pre-allocated pool for each route blueprint managed by the router specific dynamic configuration manager. This can be overriden by an annotation router.openshift.io/pool-size on an individual route.")
```
> on an individual blueprint route

You need to explain blueprint routes in the command help for this.

smarterclayton commented on pkg/cmd/infra/router/template.go:
```diff
@@ -113,12 +132,19 @@ func (o *TemplateRouter) Bind(flag *pflag.FlagSet) {
 	flag.StringVar(&o.MetricsType, "metrics-type", util.Env("ROUTER_METRICS_TYPE", ""), "Specifies the type of metrics to gather. Supports 'haproxy'.")
+	flag.StringVar(&o.ConfigManagerName, "config-manager", util.Env("ROUTER_CONFIG_MANAGER", ""), "Specifies the manager to use for dynamically configuring changes with the underlying router. Supports 'haproxy-manager'.")
+	flag.StringVar(&o.ConfigManagerConnectionInfo, "config-manager-connection-info", "", "Specifies connection information for the dynamic configuration manager.")
+	flag.DurationVar(&o.CommitInterval, "commit-interval", getIntervalFromEnv("COMMIT_INTERVAL", defaultCommitInterval), "Controls how often to commit (to the actual config) all the changes made using the router specific dynamic configuration manager.")
```
Why is this configurable?

smarterclayton replied:
I would tie this to the router reload interval unless you have some sort of real reason to make this configurable.

ramr replied:
The reason is a separate config option is because we don't want to change the reload interval to someting higher. As that means if the dynamic config manager can't bring routes online (because of certs/custom annotations etc), the reloads to bring those route would be staggered by the reload interval - not good if say that's 1 hour ... which the default for the commit interval.

smarterclayton commented on pkg/cmd/infra/router/template.go:
```diff
@@ -113,12 +132,19 @@ func (o *TemplateRouter) Bind(flag *pflag.FlagSet) {
 	flag.StringVar(&o.TemplateFile, "template", util.Env("TEMPLATE_FILE", ""), "The path to the template file to use")
 	flag.StringVar(&o.ReloadScript, "reload", util.Env("RELOAD_SCRIPT", ""), "The path to the reload script to use")
-	flag.DurationVar(&o.ReloadInterval, "interval", reloadInterval(), "Controls how often router reloads are invoked. Mutiple router reload requests are coalesced for the duration of this interval since the last reload time.")
+	flag.DurationVar(&o.ReloadInterval, "interval", getIntervalFromEnv("RELOAD_INTERVAL", defaultReloadInterval), "Controls how often router reloads are invoked. Mutiple router reload requests are coalesced for the duration of this interval since the last reload time.")
```
why did this change?

ramr replied:
I just took the reloadInterval code and made it into a reusable function so that we can call it to get the interval value for both reload and commit intervals (same code passes a different parameters).

smarterclayton commented on pkg/router/template/template_helper.go:
```diff
@@ -179,8 +195,9 @@ var helperFunctions = template.FuncMap{
 
 	"genSubdomainWildcardRegexp": genSubdomainWildcardRegexp, //generates a regular expression matching the subdomain for hosts (and paths) with a wildcard policy
-	"generateRouteRegexp":        generateRouteRegexp,        //generates a regular expression matching the route hosts (and paths)
+	"generateRouteRegexp":        GenerateRouteRegexp,        //generates a regular expression matching the route hosts (and paths)
```
Why is this exported?

smarterclayton replied:
If you are trying to reuse this, put it in its own package (`templatehelpers` maybe) and include it here.  That gets more crap out of this package.

ramr replied:
ok will do.

smarterclayton commented on pkg/router/template/router.go:
```diff
@@ -530,16 +559,168 @@ func (r *templateRouter) DeleteServiceUnit(id string) {
+		glog.V(4).Infof("Dynamically replacing endpoints for associated backend %s", backendKey)
+		if err := r.dynamicConfigManager.ReplaceRouteEndpoints(backendKey, oldEndpoints, newEndpoints, weight); err != nil {
+			// Error dynamically modifying the config, so return false to cause a reload to happen.
+			glog.Warningf("Router will reload as dynamic endpoint replacement for service id %s (backend=%s, weight=%v) failed: %v", id, backendKey, weight, err)
```
You could easily spew errors here, i'm a little concerned with this in the long term.

ramr replied:
Will change to `glog.V(4).Infof`, so that its available if someone turns up the debugging level and for all the other "Router will reload" messages.

smarterclayton commented on pkg/router/template/router.go:
```diff
@@ -530,16 +559,168 @@ func (r *templateRouter) DeleteServiceUnit(id string) {
+
+	// If no initial sync was done, don't try to dynamically add the
+	// route as we will need a reload anyway.
+	if !r.synced {
```
Why is this after registration?

ramr replied:
Because we don't add the routes if the initial sync has not been done. There's nothing in the haproxy config at that point in time (backends don't exist for blueprints as yet, etc).  Plus since there's going to be a reload and the config is going to change anyway, the configmanager is not really going to do anything here (other than registration). 
So, the registration is done always - irrespective of whether the router sync state. And that ensures that the mappings in the configmanager are correct.

smarterclayton commented:
This needs an e2e stress test.  I'd recommend the following structure:

```
for some number of iterations
  perform a random set of mutations across 2 namespaces like - add route, remove route, add service, remove service, change weight, remove endpoint, add endpoint, run conflicting host, remove conflicting host, cause a change that has an annotation (that should force a reload)
  wait up to X seconds until the router matches the state in the current namespace by checking curl (using the new route url tester)
```

I won't be comfortable merging this until I've seen a good stress test and a lot of soak time in the GCP runs.  I'd rather hold it in as a PR into 3.11 than risk any sort of regression.

smarterclayton commented:
A good start though.

ramr commented:
The base glide update PR #19158 needs rebasing (and so this PR also needs to change). 
Added support to watch blueprint routes and update the blueprints, addressed most of the review comments, added in more tests, used a boolean to enable the haproxy config manager. 
Still pending - will add an extra/specific e2e test - the existing tests do already stress this code (well with the new changes if you set an environment variable `export ROUTER_HAPROXY_CONFIG_MANAGER=true`).

knobunc commented:
/lgtm

knobunc commented:
/assign @mfojtik

ramr commented:
/retest

deads2k commented:
I heard there were some concerns about bringing in new vendored dependencies with the same pull doing the code and thought I'd clarify how the vendoring process works and why we don't need to fear it as much as godep of the past.

When people run `hack/update-deps.sh` today they are forced to managed the entire vendor tree for the product. Every update must be accepted or manually managed inside of the glide.yaml by accepting, pinning, or forking the dependency.  While that can cause some pain at the time, managing it as they come in ensures that glide's conflict and satisfaction logic remains useful to us.  Most of the time if you get updates you don't recognize, it's actually a sign that either your component or the component depending on it should move out of tree so you only have to manage one set of dependencies, not the four or so that live in origin today.  Since all updates from `hack/update-deps.sh` are managed, the delta is usually fairly small.  Glide also runs quickly and rarely crashes.

I would suggest that if your pull requires a new dependency, you bring it in at the same time as the code that uses it.  When you review the vendor bump, you're looking to make sure 
1. the content of k8s.io/* didn't change
2. openshift/* changes are *usually* ok when tests pass
3. the list of package that did change is what you expected (just your new dep and its transitives)
4. the version of another dependency didn't change.
    1. if it does change, check against kube/kube to see if kube has a preferred level.  If so, pin to the kube level with a *very* clear comment explaining it.

If you're not sure, the master team has fairly significant dependency management experience.  You can tag us in.

deads2k commented on glide.lock:
```diff
@@ -938,7 +944,7 @@ imports:
   - user/listers/user/v1
 - name: github.com/openshift/imagebuilder
-  version: 4670fc31fca6121da0dcf42d6b9b298479c1715a
+  version: bfc0aea02ce95dcfc1a83d452270dbc66933709b
```
this one is ok.

deads2k commented:
bump looks clean to me.

knobunc commented:
/lgtm

knobunc commented:
/approve

deads2k commented:
/approve

deads2k commented:
> /approve

hrm. glide.yaml may need top level.  I'll approve for the bump.

knobunc commented:
/retest

ramr commented:
@deads2k  had to rebase to upstream master - the new commit is the last `bump(*)` one:
 https://github.com/openshift/origin/pull/19073/commits/3d44247d3313dbbbe253477aa437c3fd801cae0f

Did bring in a lot, so if you can confirm that's good. Thx again.

ramr commented:
had to collapse the two `bump(*)` commits - think they were causing issues with `glide.lock`.

ramr commented:
@deads2k  hmm is this correct - the failure log for https://storage.googleapis.com/origin-ci-test/pr-logs/pull/19073/test_pull_request_origin_unit/11456/build-log.txt build failure logs shows:
```
$ git fetch https://github.com/openshift/origin.git --tags --prune
$ git fetch https://github.com/openshift/origin.git master
$ git checkout 5213e077a8ba96e2ed8afb488faa67176c7a46f8
$ git branch --force master 5213e077a8ba96e2ed8afb488faa67176c7a46f8
$ git checkout master
```
That's sha `5213e077a8ba96e2ed8afb488faa67176c7a46f8` is a commit from April 2nd.
Do we merge this PR from a commit that far off?

deads2k commented:
> @deads2k hmm is this correct - the failure log for https://storage.googleapis.com/origin-ci-test/pr-logs/pull/19073/test_pull_request_origin_unit/11456/build-log.txt build failure logs shows:
> 
> $ git fetch https://github.com/openshift/origin.git --tags --prune
> $ git fetch https://github.com/openshift/origin.git master
> $ git checkout 5213e077a8ba96e2ed8afb488faa67176c7a46f8
> $ git branch --force master 5213e077a8ba96e2ed8afb488faa67176c7a46f8
> $ git checkout master
> That's sha 5213e077a8ba96e2ed8afb488faa67176c7a46f8 is a commit from April 2nd.
> Do we merge this PR from a commit that far off?

that I do not know.  @stevekuznetsov does that look right to you?

deads2k commented:
> had to collapse the two bump(*) commits - think they were causing issues with glide.lock.

the bump still looks good. We have https://github.com/openshift/origin/pull/19969 also in queue which resolves the api changes.

knobunc commented:
/lgtm

deads2k commented:
/refresh
/skip

@ramr we renamed the job.  I bet that's an old one.

knobunc commented:
/retest

ramr commented:
/retest

knobunc commented:
@ramr Glide conflicts again in https://openshift-gce-devel.appspot.com/build/origin-ci-test/pr-logs/pull/19073/test_pull_request_origin_unit/11456/

knobunc commented:
@deads2k Any idea what is going wrong with this?

knobunc commented:
/retest

knobunc commented:
/retest

stevekuznetsov commented:
/skip

knobunc commented:
/retest

ramr commented:
/test integration
/test extended_image_ecosystem

knobunc commented:
/retest

ramr commented:
sigh, one more rebase and updated dependencies
@deads2k  there was a change to glide.lock (bump commit sha: 8681ecc03538028545d08816527e126a45f333d6)

ramr commented:
/retest

ramr commented:
/retest

ramr commented:
@deads2k  can you please reconfirm the `bump(*)` commits and approve. Thx

The bot saying a test failed is an old test - that actually threw me off a few times (I completely missed your comment about that ) before I realized a couple of days back that the build date was from April. 
D'oh!

fyi @knobunc

knobunc commented:
/retest
/lgtm

knobunc commented:
/retest
